### PR TITLE
Add "dir=auto" for input/textarea elements by default

### PR DESCRIPTION
**Pull request by wxiaoguang**

### CODE REVIEW

**Comment by wxiaoguang** [comment]
It seems fast enough for various pages.

```js
const observer = new MutationObserver((mutationList) => {
  const start = performance.now();
  for (const mutation of mutationList) {
    for (const el of mutation.addedNodes) {
      if (el.nodeType === Node.ELEMENT_NODE || el.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
        for (const child of el.querySelectorAll('input:not([dir]), textarea:not([dir])')) {
          child.setAttribute('dir', 'auto');
        }
      }
    }
  }
  console.log(`MutationObserver took ${performance.now() - start} ms.`);
});
observer.observe(document, {subtree: true, childList: true});
```


image

**Comment by silverwind** [comment]
Should we filter to exclude type `hidden` and `checkbox`? I don't see an issue adding it to them, but it might trigger obscure bugs in obscure browsers.

File: web_src/js/webcomponents/dir-auto.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    for (const el of mutation.addedNodes) {
+      if (el.nodeType === Node.ELEMENT_NODE || el.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
+        for (const child of el.querySelectorAll('input:not([dir]), textarea:not([dir])')) {
+          child.setAttribute('dir', 'auto');
+        }
+      }
+    }
+  }
+});
+observer.observe(document, {subtree: true, childList: true});
```
Comment (silverwind) [inline web_src/js/webcomponents/dir-auto.js]:
```suggestion
const excludeTypes = new Set(['hidden', 'checkbox']);

(new MutationObserver((mutationList) => {
  for (const mutation of mutationList) {
    for (const el of mutation.addedNodes) {
      if (el.nodeType === Node.ELEMENT_NODE || el.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
        for (const child of el.querySelectorAll('input:not([dir]), textarea:not([dir])')) {
          if (excludeTypes.has(child.type)) continue;
          child.setAttribute('dir', 'auto');
        }
      }
    }
  }
})).observe(document, {subtree: true, childList: true});
```

**Comment by silverwind** [comment]
Will this work on both static and js-injected content?

**Comment by wxiaoguang** [comment]
> Will this work on both static and js-injected content?

Yes, just like the `tooltip`, it works for Vue

**Comment by silverwind** [comment]
Can keep without the filter as well I guess, it will be slightly faster without.

File: web_src/js/webcomponents/dir-auto.js (hunk 1 above)
Comment (wxiaoguang) [reply web_src/js/webcomponents/dir-auto.js]: I think it needs more, see 93abd9c094d3f3e634359ec7c1d0a7f39d37c1f5

**Comment by wxiaoguang** [comment]
> Can keep without the filter as well I guess, it will be slightly faster without.

Didn't see any noticeable difference ..... I guess keeping the `:not([dir])` is good, in case some elements already have their own `dir`.

**Comment by silverwind** [comment]
I'd remove the `:not`and use `getElementsByTagName`. I don't think we will have any elements that set it, and if they do we should override it anyways.

**Comment by silverwind** [comment]
Interestingly as per [this](https://www.measurethat.net/Benchmarks/Show/3883/0/queryselectorall-vs-getelementsbytagname-iteration), `querySelectorAll` is faster in Chrome, and they are about equal in Firefox.

**Comment by silverwind** [comment]
Benchmark was flawed, [here](https://www.measurethat.net/Benchmarks/Show/27005/0/queryselectorall-vs-getelementsbytagname-iteration2#latest_results_block) is a fixed fork using `for-of`, it does show `getElementsByTagName` being around 20-30% faster in Firefox, Chrome and Safari, so I prefer we use it.

**Comment by silverwind** [comment]
OTOH, `getElementsByTagName` returns a "live" list, which is probably not what we want, so let's keep `querySelectorAll`, just remove the `:not`s.

**Comment by silverwind** [comment]
Did a few tweaks and also added some temporary benchmark logging. What worries me is that we spend 10-20ms in this function each page load:

image

**Comment by silverwind** [comment]
Imho the perf impact of this is too much. I can easily see this observer taking 1-2 seconds of time on a large diff page for example. The info mentioned [here](https://stackoverflow.com/a/39332340/808699) is very much relevant.

I think https://github.com/go-gitea/gitea/pull/26715 is the better solution that can not possibly have such a potentially huge perf impact like this has.

**Review by silverwind — CHANGES_REQUESTED** [review]
Perf impact too big

**Comment by wxiaoguang** [comment]
There are still some key points, eg: `for of` is 3-5 times slows than `for(i++)` (tested by the empty loops).

**Comment by wxiaoguang** [comment]
And another point is: it doesn't need to be executed in the `<head>`, it could also be executed "on DOM loaded"

File: web_src/js/webcomponents/dir-auto.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+const includeNodeTypes = new Set([Node.ELEMENT_NODE, Node.DOCUMENT_FRAGMENT_NODE]);
+
+let timeSpent = 0;
+
+const observer = new MutationObserver((mutationList) => {
+  const start = performance.now();
+  for (const mutation of mutationList) {
+    for (const el of mutation.addedNodes) {
+      if (!includeNodeTypes.has(el.nodeType)) continue;
+      if (!el.querySelector('input, textarea')) continue;
```
Comment (wxiaoguang) [inline web_src/js/webcomponents/dir-auto.js]: This is a no-op (and makes the situation worse)

**Comment by wxiaoguang** [comment]
Fully rewritten, now it takes 1.5ms on a PR page with `+6966 -3615`

File: web_src/js/webcomponents/webcomponents.js [hunk 3]
```diff
@@ -3,3 +3,4 @@ import './polyfill.js';
 
 import '@github/relative-time-element';
 import './GiteaOriginUrl.js';
+
```
Comment (silverwind) [inline web_src/js/webcomponents/webcomponents.js]:
```suggestion
```

File: web_src/js/modules/dirauto.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// for performance considerations, it only uses performant syntax
+
+function attachDirAuto(el) {
+  if (el.type !== 'hidden' &&
+    el.type !== 'checkbox' &&
+    el.type !== 'radio' &&
+    el.type !== 'range' &&
+    el.type !== 'color') {
```
Comment (silverwind) [inline web_src/js/modules/dirauto.js]: a `Set` should be faster than this.

**Comment by silverwind** [comment]
I guess ideally we should only ever install one async `MutationObserver` and do both the tippy and this modification in it.

File: web_src/js/modules/dirauto.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+export function initDirAuto() {
+  let timeSpent = 0;
+
+  const observer = new MutationObserver((mutationList) => {
+    const start = performance.now();
+    for (let mutationIdx = 0; mutationIdx < mutationList.length; mutationIdx++) {
+      const mutation = mutationList[mutationIdx];
+      for (let addedNodeIdx = 0; addedNodeIdx < mutation.addedNodes.length; addedNodeIdx++) {
+        const addedNode = mutation.addedNodes[addedNodeIdx];
+        if (addedNode.nodeType !== Node.ELEMENT_NODE && addedNode.nodeType !== Node.DOCUMENT_FRAGMENT_NODE) continue;
```
Comment (silverwind) [inline web_src/js/modules/dirauto.js:21]: Also would keep the `Set` here which is faster.

**Comment by wxiaoguang** [comment]
> a Set should be faster than this.
> ...
>  Also would keep the Set here which is faster.

That's impossible. Benchmark says my code is faster:


```js

const excludeInputTypesSet = new Set(['hidden', 'checkbox', 'radio', 'range', 'color']);

function includeInputTypes(t) {
  return (t !== 'hidden' &&
    t !== 'checkbox' &&
    t !== 'radio' &&
    t !== 'range' &&
    t !== 'color');
}

for (let repeat = 0; repeat < 3; repeat++) {
  const count = 1000000;
  let tmp = 0;
  let start = performance.now();
  for (let i = 0; i < count; i++) {
    if (!excludeInputTypesSet.has('text')) tmp++;
    if (!excludeInputTypesSet.has('hidden')) tmp++;
    if (!excludeInputTypesSet.has('color')) tmp++;
  }
  const d1 = performance.now() - start;

  start = performance.now();
  for (let i = 0; i < count; i++) {
    if (includeInputTypes('text')) tmp--;
    if (includeInputTypes('hidden')) tmp--;
    if (includeInputTypes('color')) tmp--;
  }
  const d2 = performance.now() - start;
  if (tmp !== 0) throw new Error('mismatched');

  console.log(`by set: ${d1}, by equal: ${d2}`);
}
```


```
by set: 13.899999976158142, by equal: 2.100000023841858
dirauto.js:76 by set: 14.900000095367432, by equal: 0.5999999046325684
dirauto.js:76 by set: 12.100000023841858, by equal: 1.399999976158142
```

**Comment by silverwind** [comment]
Strange. BTW, a `switch` statement is even faster than an `if-else` chain.

**Comment by silverwind** [comment]
Tested with `switch`, seems about equal here.

**Comment by puni9869** [comment]
> ```js
> function includeInputTypes(t) {
>   switch (t) {
>     case 'hidden':
>     case 'checkbox':
>     case 'radio':
>     case 'range':
>     case 'color':
>       return false;
>   }
>   return true;
> }
> ```

Please forgive me, I have better version in `wasm`.

```asm
section .data
    inputTypes db 'hidden checkbox radio range color', 0
    falseMsg db 'false', 0
    trueMsg db 'true', 0

section .bss
    userChoice resb 16

section .text
    global includeInputTypes
includeInputTypes:
    mov rsi, inputTypes  ; Load the address of inputTypes into rsi

.loop:
    mov rdi, userChoice  ; Load the address of userChoice into rdi
    call readInput       ; Assume a readInput function to read user input

    ; Compare the user input with each input type
    mov rcx, rsi         ; Store the address of inputTypes in rcx
    mov al, byte [rdi]   ; Load the first byte of userChoice into al
    cmp byte [rcx], 0    ; Check if the current character in inputTypes is null terminator
    je .notIncluded      ; If yes, jump to notIncluded

.checkType:
    cmp al, byte [rcx]   ; Compare the user input with the current character in inputTypes
    je .included         ; If equal, jump to included
    inc rcx              ; Move to the next character in inputTypes
    jmp .checkType       ; Repeat the loop for the next character

.included:
    mov rax, falseMsg    ; Load the address of falseMsg into rax
    ret

.notIncluded:
    mov rax, trueMsg     ; Load the address of trueMsg into rax
    ret

readInput:
    ; Assume implementation for reading user input
    ret

```

PS: it was just for fun, intention was not to hurt anyone sentiments

**Comment by wxiaoguang** [comment]
> Tested with `switch`, seems about equal here.

Trust the browser's optimization and JIT

**Comment by silverwind** [comment]
FYI I have not yet decided which of the two methods to go for. Will evaluate soon.

**Comment by silverwind** [comment]
Switched to set, it [is faster](https://perf.link/#eyJpZCI6IjM0bXJ1NDdpdGRmIiwidGl0bGUiOiIiLCJiZWZvcmUiOiJjb25zdCBleGNsdWRlVHlwZXMgPSBuZXcgU2V0KFsnaGlkZGVuJywgJ2NoZWNrYm94JywgJ3JhZGlvJywgJ3JhbmdlJywgJ2NvbG9yJ10pO1xuY29uc3QgZWwgPSB7fSIsInRlc3RzIjpbeyJuYW1lIjoiIiwiY29kZSI6ImlmICghZXhjbHVkZVR5cGVzLmhhcyhlbC50eXBlKSkge1xuICBlbC5kaXIgPSAnYXV0byc7XG59IiwicnVucyI6WzY3ODAwMCw0OTUwMDAsNjM0MDAwLDg1ODAwMCw0OTUwMDAsMzk1MDAwMCwyMzkwMDAwLDUyNDAwMCwzNTA2MDAwLDYyOTAwMCw1MTUwMDAsMjk1MDAwLDUzNzAwMCwxNjk1MDAwLDIzOTIwMDAsMjI3MDAwLDQ3NjEwMDAsMzY5MDAwLDYzNjAwMCw2MjcwMDAsMjE1MDAwLDQ3MDAwLDI4MTkwMDAsMjI5NzAwMCwxMzI1MDAwLDExMjAwMCwyMzEyMDAwLDU2NzAwMCwzOTE4MDAwLDQ5OTAwMCwyOTUyMDAwLDk5NTAwMCw0NDcwMDAsNDk1MDAwLDQxMDMwMDAsOTI3MDAwLDE1OTcwMDAsNjgzMDAwLDE5ODgwMDAsNjA2MDAwLDI4MjgwMDAsNTEzOTAwMCwxMTAyMDAwLDQyMDQwMDAsMzUwNTAwMCw1ODIwMDAsNDgzNzAwMCw2NTUwMDAsMTY1NjAwMCwyMjEwMDAsNzcwMDAwLDcxMDAwMCw1NjkwMDAsNjg1MDAwLDYwNDAwMCwxNTg1MDAwLDI3NjEwMDAsMzYzOTAwMCw3MDAwMDAsNjgxMDAwLDM0NDAwMCw4MzQwMDAsNjI2MDAwLDE3NDkwMDAsNzE1MDAwLDM2NDMwMDAsMzE4MzAwMCwzMjc5MDAwLDE5MTMwMDAsNjc5MDAwLDQzMzcwMDAsMzkzMDAwLDEwMDAsMzQ1MzAwMCwyODQzMDAwLDY4NzAwMCwxNzg2MDAwLDQyMjkwMDAsNTI2MDAwLDYyOTUwMDAsMjMzMDAwMCw2MDQwMDAwLDY0NTAwMCw2NjYwMDAsMjE0NDAwMCw5MTIwMDAsMjQ2MDAwLDU3NDAwMCwzODI2MDAwLDkyMTAwMCwxMDI1MDAwLDUwODAwMCw5MjcwMDAsNjczMDAwLDIzOTAwMCwyNjkwMDAsNjE1MDAwLDUwNjAwMCw2NTcwMDAsMzU1MjAwMF0sIm9wcyI6MTYxMzQwMH0seyJuYW1lIjoiIiwiY29kZSI6IiAgaWYgKGVsLnR5cGUgIT09ICdoaWRkZW4nICYmXG4gICAgZWwudHlwZSAhPT0gJ2NoZWNrYm94JyAmJlxuICAgIGVsLnR5cGUgIT09ICdyYWRpbycgJiZcbiAgICBlbC50eXBlICE9PSAncmFuZ2UnICYmXG4gICAgZWwudHlwZSAhPT0gJ2NvbG9yJykge1xuICAgIGVsLmRpciA9ICdhdXRvJztcbiAgfSIsInJ1bnMiOls2MDAwMCw0NzIwMDAsNDYwMDAwLDUwMTAwMCw1NzQwMDAsMzEwMDAwLDEwMzMwMDAsNzAwMDAsMTAwMCw0NDYwMDAsMTUxNTAwMCw1NzcwMDAsMzg3MDAwLDQ5NTAwMCw5NjgwMDAsMTk3MzAwMCwyOTcxMDAwLDI5MzAwMCw1NzEwMDAsNTI3MDAwLDU5OTAwMCwxOTYzMDAwLDI0NzAwMDAsMTQxNzAwMCwxMzA3MDAwLDE1ODUwMDAsNTAwMCwzNDMwMDAsMzk3MDAwLDUxNzAwMCwzNTk2MDAwLDU2OTAwMCw1NTAwMDAsMjkwMDAwLDMzNjgwMDAsNTAwMDAwLDE0NDIwMDAsNDEyMDAwLDE0MjUwMDAsMjIxMDAwLDEwMjgwMDAsMjQ5NDAwMCwzNjYwMDAsMzgzMDAwLDMwMjAwMCwzNjEwMDAsMjAyMDAwMCw1NDcwMDAsMTM4MzAwMCwxNzAyMDAwLDUwMDAwMCwxNTY2MDAwLDU1OTAwMCw1MTkwMDAsMjM2MDAwLDU2MzAwMCw4OTgwMDAsMjQ4MTAwMCw0OTAwMDAsNDQ1MDAwLDE2OTAwMCwzMzgzMDAwLDMxNDAwMCw1NDgwMDAsMTIyMDAwLDAsMTk4MzAwMCwyNjY3MDAwLDEyOTAwMDAsNDg4MDAwLDE0NDAwMCwzNDYwMDAsMjcxNjAwMCwyNjAyMDAwLDE5NDgwMDAsNDQ0MDAwLDEyMzAwMDAsMjg5NDAwMCwxNzkwMDAsNTI3MTAwMCwxNjAwMDAwLDMyMTAwMDAsNDA0MDAwLDI1MTAwMCw0NzAwMCwyNzg2MDAwLDkwMDAwLDUzMTAwMCwxMDcxMDAwLDUzOTAwMCwyNzkwMDAwLDEwNjAwMCwzNTEwMDAsMjkwMDAwLDIzMTcwMDAsMjIwMDAwLDM4ODAwMCwxNDAwMCw0MTgwMDAsMTc4MDAwXSwib3BzIjoxMDIzNjIwfV0sInVwZGF0ZWQiOiIyMDIzLTA5LTA2VDIwOjE1OjM3LjY1MVoifQ%3D%3D).

**Comment by silverwind** [comment]
Another small gain made with [length caching](https://perf.link/#eyJpZCI6Im41ZjZoc3pqNHRkIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGRhdGEgPSBbLi4uQXJyYXkoMTAwMCkua2V5cygpXSIsInRlc3RzIjpbeyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImNvbnN0IGxlbiA9IGRhdGEubGVuZ3RoO1xuZm9yKGxldCBpID0gMDsgaSA8IGxlbjsgaSsrKTsiLCJydW5zIjpbMTAwMDAsNTAwMDAsMzkwMDAsMjMwMDAsMTAwMCwyMzAwMCw0NjAwMCwyNjAwMCwxNDgwMDAsNzMwMDAsMzYwMDAsMjcwMDAsMTkwMDAsMTYwMDAsMTgwMDAsMzkwMDAsOTcwMDAsNzgwMDAsMzAwMDAsNzcwMDAsNTcwMDAsMTIxMDAwLDIwMDAwLDYxMDAwLDIzMDAwLDUxMDAwLDQyMDAwLDEwMDAsMTMwMDAsNDYwMDAsODMwMDAsMTAwMDAsMjMwMDAsMjkwMDAsMjMwMDAsMTE0MDAwLDU5MDAwLDM1MDAwLDgxMDAwLDM5MDAwLDcwMDAwLDMyMDAwLDExOTAwMCwxMDAwMCw3NTAwMCw3MDAwLDg3MDAwLDQ5MDAwLDIyMDAwLDgwMDAsMjQwMDAsNTkwMDAsMjMwMDAsNjAwMDAsMTAwMCwzMDAwMCw0MzAwMCw0MjAwMCw3NzAwMCw0ODAwMCw1NDAwMCwzMjAwMCw0MzAwMCw0NDAwMCwyOTAwMCw0NjAwMCwyMzAwMCwzMDAwLDYzMDAwLDIxMDAwLDg2MDAwLDEwMDAsNjcwMDAsMTAwMCw0OTAwMCw5MjAwMCw5NzAwMCwyNzAwMCwyMzAwMCwyMDAwMCwxMDIwMDAsMzEwMDAsMjUwMDAsNDAwMCwxNjAwMCwxODAwMCwyMzAwMCwyMDAwLDIzMDAwLDgwMDAsMTIwMDAsNjEwMDAsMzYwMDAsNTEwMDAsMTAwMCwyMjAwMCw0MjAwMCwyNDAwMCwxNjAwMCwxMzAwMF0sIm9wcyI6NDA0NDB9LHsibmFtZSI6IkZpbmQgaXRlbSA4MDAiLCJjb2RlIjoiZm9yKGxldCBpID0gMDsgaSA8IGRhdGEubGVuZ3RoOyBpKyspOyIsInJ1bnMiOlsxMTAwMCwxNjAwMCw0MDAwLDIxMDAwLDEwMDAwLDM0MDAwLDI5MDAwLDE1MDAwLDQxMDAwLDMxMDAwLDI0MDAwLDI0MDAwLDU1MDAwLDcxMDAwLDQyMDAwLDMwMDAsNTcwMDAsMTgwMDAsMTAwMDAsMTUwMDAsMzYwMDAsMjcwMDAsMjkwMDAsMTQwMDAsNTAwMCwyNjAwMCw1MzAwMCwyMTAwMCwxMTAwMCwxOTAwMCwxMjAwMCw1MDAwMCw2MDAwLDUzMDAwLDE1MDAwLDMzMDAwLDIzMDAwLDkwMDAsMzkwMDAsMjgwMDAsMjcwMDAsNjAwMCwzMjAwMCw2MDAwLDE4MDAwLDMwMDAsMTYwMDAsNDAwMDAsNjMwMDAsMzQwMDAsNDAwMCwyMTAwMCwyOTAwMCwxMjAwMCwxOTAwMCwxMDAwLDM2MDAwLDgwMDAsNDIwMDAsMTUwMDAsMTUwMDAsNDMwMDAsNDAwMDAsMjAwMDAsODAwMCwyNzAwMCw0OTAwMCw1NjAwMCwxODAwMCw2MTAwMCwxNTAwMCwxNTAwMCwyNTAwMCwxMDAwMCw1NTAwMCwyODAwMCw0MjAwMCwxMDAwLDQ5MDAwLDMwMDAwLDE2MDAwLDI2MDAwLDEwMDAsMTYwMDAsMTUwMDAsMzkwMDAsNTQwMDAsMjAwMDAsNjAwMCwxMTAwMCwxNDAwMCwxNTAwMCw0ODAwMCwxNjAwMCwxNTAwMCwyNTAwMCwzMDAwMCw3MDAwLDQ1MDAwLDcwMDBdLCJvcHMiOjI1MDUwfV0sInVwZGF0ZWQiOiIyMDIzLTA5LTA2VDIwOjI1OjEwLjA1MloifQ%3D%3D).

**Review by silverwind — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> Switched to Set, it is faster at least with 5 elements.

I am sure the "perf.link" is buggy. It shows my code is faster too. While such minor performance doesn't bother me, either is fine.

image

**Review by puni9869 — APPROVED** [review]

**Comment by puni9869** [comment]
Let's roll 🚀

**Comment by silverwind** [comment]
> I am sure the "perf.link" is somewhat/somehow buggy. It shows my code is faster too. While such minor performance doesn't bother me, either is fine.

At least Chrome and Safari show a significant improvement for Set. In Firefox it is closer for me too, but still 5-10% faster.

**Comment by wxiaoguang** [comment]
> > I am sure the "perf.link" is somewhat/somehow buggy. It shows my code is faster too. While such minor performance doesn't bother me, either is fine.
> 
> At least Chrome and Safari show a significant improvement for Set. In Firefox it is closer for me too, but still 5-10% faster.

How do your browser think about "https://github.com/go-gitea/gitea/pull/26735#issuecomment-1694260205" ? On my side, Set is significantly slower.

The "perf.link" is just a blackbox, I don't fully trust it for some critical tasks.

**Comment by silverwind** [comment]
> How do your browser think about "https://github.com/go-gitea/gitea/pull/26735#issuecomment-1694260205"

Firefox
```
by set: 2.589832999976352, by equal: 2.4636940001510084
by set: 2.9552920000860468, by equal: 0.8551479999441653
by set: 1.521970999892801, by equal: 0.7742269999580458
```

Chrome
```
by set: 42.900000005960464, by equal: 5.200000002980232
by set: 28.200000002980232, by equal: 1.1000000089406967
by set: 16.399999991059303, by equal: 0.5
```

**Comment by wxiaoguang** [comment]
> > How do your browser think about "[#26735 (comment)](https://github.com/go-gitea/gitea/pull/26735#issuecomment-1694260205)"
>
> Chrome
> 
> ```
> by set: 42.900000005960464, by equal: 5.200000002980232
> by set: 28.200000002980232, by equal: 1.1000000089406967
> by set: 16.399999991059303, by equal: 0.5
> ```

That's a huge different result from perf.link. I think our manually written test code is more like the real case we are facing.

**Comment by silverwind** [comment]
Switched it back

**Pull request merged**